crates/compiler/serialize/src/bytes.rs

```diff
@@ -0,0 +1,381 @@
+use std::{
+    borrow::Borrow,
+    io::{self, Write},
+};
+
+use roc_collections::{MutMap, VecMap};
+
+pub fn serialize_slice<T: Copy>(
+    slice: &[T],
+    writer: &mut impl Write,
+    written: usize,
+) -> io::Result<usize> {
+    let alignment = std::mem::align_of::<T>();
+    let padding_bytes = round_to_multiple_of(written, alignment) - written;
+
+    for _ in 0..padding_bytes {
+        writer.write_all(&[0])?;
+    }
+
+    let bytes_slice = unsafe { slice_as_bytes(slice) };
+    writer.write_all(bytes_slice)?;
+
+    Ok(written + padding_bytes + bytes_slice.len())
+}
+
+pub fn deserialize_slice<T: Copy>(bytes: &[u8], length: usize, mut offset: usize) -> (&[T], usize) {
+    let alignment = std::mem::align_of::<T>();
+    let size = std::mem::size_of::<T>();
+
+    offset = round_to_multiple_of(offset, alignment);
+
+    let byte_length = length * size;
+    let byte_slice = &bytes[offset..][..byte_length];
+
+    let slice = unsafe { std::slice::from_raw_parts(byte_slice.as_ptr() as *const T, length) };
+
+    (slice, offset + byte_length)
+}
+
+pub fn deserialize_vec<T: Clone + Copy>(
+    bytes: &[u8],
+    length: usize,
+    offset: usize,
+) -> (Vec<T>, usize) {
+    let (slice, offset) = deserialize_slice(bytes, length, offset);
+    (slice.to_vec(), offset)
+}
+
+#[derive(Copy, Clone)]
+struct VecSlice<T> {
+    pub start: u32,
+    pub length: u16,
+    _marker: std::marker::PhantomData<T>,
+}
+
+impl<T> VecSlice<T> {
+    const fn len(&self) -> usize {
+        self.length as usize
+    }
+
+    fn indices(&self) -> std::ops::Range<usize> {
+        self.start as usize..(self.start as usize + self.length as usize)
+    }
+
+    fn extend_new(vec: &mut Vec<T>, it: impl IntoIterator<Item = T>) -> Self {
+        let start = vec.len();
+
+        vec.extend(it);
+
+        let end = vec.len();
```

Comment:
would https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.count work here?

Reply:
No, because that consumes the iterator